### ISSUE TITLE
Patch does not run if specific other patch exists

### ISSUE THREAD

**Issue by SpikeHimself** [BepInEx/HarmonyX#71]
Apologies in advance for the vague description. This is a mind boggling issue for me (I'm new to this) and I'm not even sure if Harmony is at fault. I'm hoping someone can point me in the right direction with this.

Consider this pseudo code:

```
void DoThings()
{
    OneThing();
    AnotherThing();
}
```

If I patch `OneThing`, that works as expected. Prefix, postfix, finalizer, all work fine.

Now, if I also patch `DoThings`, *that* works, but any patch for `OneThing` will now be skipped.  That is to say, the debug logs say that the method did get patched, but *the code does not run*. It does not matter if it is a prefix, postfix, or finalizer. None of them run. 

In fact, if the patch on `DoThings` is empty, like so:

```
void Prefix() { }
```

Obviously it will now not do anything, and yet it *still* breaks the other patch.

Is this expected behaviour? 

The context in which I discovered this is the Unity game Valheim, and the bug (if it is one) happens in a  mod I wrote, but was initially caused by *another mod*. Meaning, my mod had patched `OneThing` and that no longer worked because another mod had patched `DoThings`. 

How can I debug this and figure out what's actually going on "behind the scenes"?

**Comment by ManlyMarco** [comment]
Try putting [HarmonyDebug] attributes on your outer prefix and see what the IL comes out like.

**Comment by SpikeHimself** [comment]
The plot thickens..

If I change the order in which I write the patches, **both work**.

Method 1, only the outer patch is called:

```c#
    // this is the outer method
    [HarmonyPatch(typeof(Player), nameof(Player.PlacePiece))]
    static class Player_PlacePiece
    {
        static void Postfix()
        {
            // This runs
            Log.Debug("Player.PlacePiece.Postfix");
        }
    }

    // this is the inner method
    [HarmonyPatch(typeof(WearNTear), nameof(WearNTear.OnPlaced))]
    static class WearNTear_OnPlaced
    {
        static void Postfix()
        {
            // This does not run
            Log.Debug("WearNTear.OnPlaced.Postfix");
        }
    }
```

..but the other way around, both run:

```c#
    // this is the inner method
    [HarmonyPatch(typeof(WearNTear), nameof(WearNTear.OnPlaced))]
    static class WearNTear_OnPlaced
    {
        static void Postfix()
        {
            // This runs!
            Log.Debug("WearNTear.OnPlaced.Postfix");
        }
    }

    // this is the outer method
    [HarmonyPatch(typeof(Player), nameof(Player.PlacePiece))]
    static class Player_PlacePiece
    {
        static void Postfix()
        {
            // This runs
            Log.Debug("Player.PlacePiece.Postfix");
        }
    }
```` 

Is there a way I can control which patch is applied first, bearing in mind that the outer patch might come from another mod that I have no control over?

**Comment by ManlyMarco** [comment]
Can you try the same but with type-wide patching instead of assembly-wide? It sounds like the outer patch still points to the original method instead of the patched method, somehow.

**Comment by SpikeHimself** [comment]
Okay, I now run this:

```c#
patcher.PatchAll(typeof(Player_PlacePiece));
patcher.PatchAll(typeof(WearNTear_OnPlaced));
```

instead of

```c#
patcher.PatchAll();
```

Is that what you meant?  The outcome is the same - the outer patch runs, the inner patch does not.

**Comment by ManlyMarco** [comment]
Thanks, @ghorsington will have to take a look at it.

**Comment by Windows10CE** [comment]
This sounds like it's just an inlining issue, if you patch the outer first, the method gets compiled before the inner patch can be applied to disable inlining for the inner method. Inlined methods effectively aren't patched, because the JIT uses the original IL to inline the method. If you patch the inner first, MonoMod disables inlining for the method permanently, meaning the inner method will no longer be inlined, and both patches will run.

**Comment by Windows10CE** [comment]
Easiest way? Get your mod loaded before the other mod, either by getting the other mod to add a SoftDependency on your mod, or having your mod get sorted higher alphabetically.

Alternatively, you could write a preloader that adds the NoInlining flag to the inner method before the assembly even loads.

**Comment by SpikeHimself** [comment]
The more I think about this, the less sense it makes in my head. Why does the inner method get inlined when a harmony patch exists on the outer method, but not if it doesn't?

Getting other mods to add a soft dependency on my mod would be an impossible task. There are a plethora of mods that patch the outer method. It seems that in a game where you can build stuff, lots of mods are interested in the build event for some reason! :P Renaming my mod isn't great either as I would have to abandon the current mod and make a new one, risking my sanity as well as a loss in endorsements and users. Writing a preloader just because I want to know when a specific item is being built in a game seems like insane overkill, I'd like to avoid that approach at all costs.

I did write a silly work-around which checks the result of the inner method after a small delay, and in that way I am able to "emulate" the circumstances under which the inner patch  "would have" run. It's very ugly but I guess it'll have to do.

The only other thing for it, I suppose, would be to write "known issue" in the comment above my patch and move on with life.

**Comment by ZenDragonX** [comment]
I am also experiencing this problem.  A year later.  Same area of the code too.  Player.PlacePiece and WearNTear.OnPlaced.

**Comment by SpikeHimself** [comment]
> I am also experiencing this problem. A year later. Same area of the code too. Player.PlacePiece and WearNTear.OnPlaced.

I don't think this is going to be fixable by Harmony, at least not in the short term. The bug is created by the compiler's inlining, and so the function being patched no longer exists when the time comes for Harmony to apply its patches.

Working around it seems the only way to cope with it for now.

**Comment by myocytebd** [comment]
> This sounds like it's just an inlining issue, if you patch the outer first, the method gets compiled before the inner patch can be applied to disable inlining for the inner method. Inlined methods effectively aren't patched, because the JIT uses the original IL to inline the method. If you patch the inner first, MonoMod disables inlining for the method permanently, meaning the inner method will no longer be inlined, and both patches will run.

This sounds a bit weird to me.
- Do you mean that HarmonyX patching on method synchronously (or immediately) trigger JIT of that method?
- What is the point for MonoMod to disable inlining for that method?
  It doesn't make sense to me: the method IL is already updated, and further inlining should use the updated IL?

BTW how does Unity Mono do JIT?
Does it only do baseline compilation at load time (or method 1st time)? Or does it do further (non-deterministic) optimizing compilation based on profiling?

**Comment by nike4613** [comment]
> @SpikeHimself: The more I think about this, the less sense it makes in my head. Why does the inner method get inlined when a harmony patch exists on the outer method, but not if it doesn't?

The TLDR is that MonoMod forcibly compiles methods with ILHooks (so anything touched by HarmonyX) so that any IL issues throw *at patch application* and MonoMod can roll back the broken patch. Otherwise, you're likely patching long before the method is first called, and therefore compiled.

Thus, if you *don't* patch outer:
- You patch inner
- Some time *much* later, outer is compiled, and since inner was patched, inner is not inlined.

Note that if you *didn't* patch inner, inner likely would have been inlined into outer.

If you *do* patch outer, and do so before inner:
- You patch outer
  - As part of patch application, outer is compiled, and inner is inlined
- You patch inner
- Much later, `outer` is invoked, and does not invoke the patched inner because inner was inlined. 

> @myocytebd: Do you mean that HarmonyX patching on method synchronously (or immediately) trigger JIT of that method?

Yes, vis-a-vis MonoMod's ILHooks. See above.

> What is the point for MonoMod to disable inlining for a patched method?
> It doesn't make sense to me: the method IL is already updated, and further inlining should use the updated IL?

We can't actually replace IL. What we do is generate a *new* method with the new IL, then do some tricks to make sure that calls to the old method instead end up at the new method (by writing in some special native instructions at the compiled method's location). We can't update inlined methods this way, for obvious reasons.

The runtime does (under certain circumstances) actually track where methods were inlined, but from in-process managed code, we both can't see that, and also can't tell the runtime to recompile methods that inlined a method we changed.

**Comment by ManlyMarco** [comment]
That's because debug mode turns off inlining and other optimizations. Obviously there is a performance hit, in some games it can be very high.
